File: tests/test_commandcode_and_argument_aliases.py
#!/usr/bin/env python3
"""
tests/test_commandcode_and_argument_aliases.py
Verifies:
1. Argument normalization and alias handling in server.py and file_tools_router.
2. Token-compact formatting of project context and dependency filtering in persona_memory.py.
3. Command Code scratch paths allowability.
4. CLI help flag and subcommand parity across all commands.
"""

import os
import sys
import json
import unittest
import subprocess

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))
import server
import persona_memory

class TestCommandCodeAndAliases(unittest.TestCase):
    def test_argument_alias_normalization_in_server(self):
        args = {
            "FilePath": "/path/to/file.txt",
            "StartLine": 5,
            "EndLine": 20
        }
        # Validate that _validate_manifest_arguments normalizes aliases into snake_case
        server._validate_manifest_arguments("read_file_range", args)
        self.assertIn("file_path", args)
        self.assertEqual(args["start_line"], 5)
        self.assertEqual(args["end_line"], 20)

    def test_project_stack_dependency_filtering(self):
        # Create a mock package.json with many dependencies
        mock_pkg = {
            "name": "test-app",
            "dependencies": {
                f"dep-{i}": "1.0.0" for i in range(50)
            },
            "devDependencies": {
                "next": "15.0.0",
                "tailwindcss": "4.0.0",
                "@tailwindcss/postcss": "4.0.0"
            }
        }
        import tempfile
        with tempfile.TemporaryDirectory() as tmp_dir:
            pkg_file = os.path.join(tmp_dir, "package.json")
            with open(pkg_file, "w") as f:
                json.dump(mock_pkg, f)

            stack = persona_memory.detect_project_stack(tmp_dir)
            self.assertIn("Next.js", stack["framework"])
            self.assertIn("Tailwind CSS v4", stack["styling"])
            # Dependencies array must be bounded (<= 15 items)
            self.assertLessEqual(len(stack["dependencies"]), 15)

    def test_format_project_context_compactness(self):
        profile = {
            "project_name": "ecommerce-frontend",
            "project_path": "/home/user/workspace/ecommerce",
            "framework": "Next.js 15.0",
            "styling": "Tailwind CSS v4 (@theme directives)",
            "package_manager": "pnpm",
            "context_summary": "Enforce strict TypeScript, Geist font, zero emojis, 10 light-mode themes, and pnpm exclusively across all components and API routes."
        }
        mems = [
            {"memory_type": "rule", "content": "Always validate forms with Zod and server actions."},
            {"memory_type": "architecture", "content": "Keep page-level routes in app/ directory with dynamic metadata."}
        ]

        formatted_normal = persona_memory.format_project_context_for_prompt(profile, mems, compact=False)
        formatted_compact = persona_memory.format_project_context_for_prompt(profile, mems, compact=True)

        self.assertIn("Persistent Project Context", formatted_normal)
        self.assertIn("Auto-Compacted", formatted_compact)
        self.assertLess(len(formatted_compact), len(formatted_normal))
        # Ensure compact block is under 350 characters
        self.assertLess(len(formatted_compact), 350)

    def test_cli_help_flags(self):
        cli_path = os.path.join(os.path.dirname(__file__), '..', 'bin', 'cli.js')
        commands_to_test = [
            ["init", "--help"],
            ["migrate", "--help"],
            ["test", "--help"],
            ["status", "--help"],
            ["savings", "--help"],
            ["doctor", "--help"],
            ["uninstall", "--help"],
            ["version", "--help"],
            ["upgrade", "--help"],
            ["skill", "--help"],
            ["agent", "--help"],
            ["models", "--help"],
            ["data", "--help"],
            ["project", "--help"],
            ["bridge", "--help"],
            ["help"]
        ]
        for cmd_args in commands_to_test:
            run = subprocess.run(["node", cli_path] + cmd_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            self.assertEqual(run.returncode, 0, f"Command 'konoha {' '.join(cmd_args)}' failed with code {run.returncode}: {run.stderr}")
            self.assertTrue(len(run.stdout) > 20, f"Command 'konoha {' '.join(cmd_args)}' returned empty output")

if __name__ == '__main__':
    unittest.main()

File: tests/test_commandcode_and_argument_aliases.js
#!/usr/bin/env node
'use strict';

/**
 * tests/test_commandcode_and_argument_aliases.js
 * Verifies:
 * 1. Argument normalization and alias handling in server and file_tools_router.
 * 2. Token-compact formatting of project context and dependency filtering in persona_memory.
 * 3. CLI help flag and subcommand parity across all commands.
 */

const assert = require('assert');
const fs = require('fs');
const os = require('os');
const path = require('path');
const { execFileSync } = require('child_process');
const server = require('../src/server');
const personaMemory = require('../src/persona_memory');

const REPO_ROOT = path.resolve(__dirname, '..');
const CLI_PATH = path.join(REPO_ROOT, 'bin', 'cli.js');

async function run() {
  console.log('Running test_commandcode_and_argument_aliases tests...');

  // 1. Argument alias normalization in server
  const args = {
    FilePath: '/path/to/file.txt',
    StartLine: 5,
    EndLine: 20
  };
  server._validate_manifest_arguments('read_file_range', args);
  assert.ok('file_path' in args, 'file_path must be in args');
  assert.strictEqual(args.start_line, 5);
  assert.strictEqual(args.end_line, 20);
  console.log('✓ Argument alias normalization passed');

  // 2. Project stack dependency filtering
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cmdcode_dep_'));
  try {
    const mockPkg = {
      name: 'test-app',
      dependencies: {},
      devDependencies: {
        next: '15.0.0',
        tailwindcss: '4.0.0',
        '@tailwindcss/postcss': '4.0.0'
      }
    };
    for (let i = 0; i < 50; i++) {
      mockPkg.dependencies[`dep-${i}`] = '1.0.0';
    }
    fs.writeFileSync(path.join(tmpDir, 'package.json'), JSON.stringify(mockPkg));

    const stack = personaMemory.detectProjectStack(tmpDir);
    assert.ok(stack.framework.includes('Next.js'));
    assert.ok(stack.styling.includes('Tailwind CSS v4'));
    assert.ok(stack.dependencies.length <= 15, `Dependencies array should be <= 15 items, got ${stack.dependencies.length}`);
    console.log('✓ Project stack dependency filtering passed');
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }

  // 3. Format project context compactness
  const profile = {
    project_name: 'ecommerce-frontend',
    project_path: '/home/user/workspace/ecommerce',
    framework: 'Next.js 15.0',
    styling: 'Tailwind CSS v4 (@theme directives)',
    package_manager: 'pnpm',
    context_summary: 'Enforce strict TypeScript, Geist font, zero emojis, 10 light-mode themes, and pnpm exclusively across all components and API routes.'
  };
  const mems = [
    { memory_type: 'rule', content: 'Always validate forms with Zod and server actions.' },
    { memory_type: 'architecture', content: 'Keep page-level routes in app/ directory with dynamic metadata.' }
  ];

  const formattedNormal = personaMemory.formatProjectContextForPrompt(profile, mems, 2, false);
  const formattedCompact = personaMemory.formatProjectContextForPrompt(profile, mems, 2, true);

  assert.ok(formattedNormal.includes('Persistent Project Context'));
  assert.ok(formattedCompact.includes('Auto-Compacted'));
  assert.ok(formattedCompact.length < formattedNormal.length);
  assert.ok(formattedCompact.length < 350, `Compact block length should be < 350, got ${formattedCompact.length}`);
  console.log('✓ Format project context compactness passed');

  // 4. CLI help flags
  const commandsToTest = [
    ['init', '--help'],
    ['migrate', '--help'],
    ['test', '--help'],
    ['status', '--help'],
    ['savings', '--help'],
    ['doctor', '--help'],
    ['uninstall', '--help'],
    ['version', '--help'],
    ['upgrade', '--help'],
    ['skill', '--help'],
    ['agent', '--help'],
    ['models', '--help'],
    ['data', '--help'],
    ['project', '--help'],
    ['bridge', '--help'],
    ['help']
  ];

  for (const cmdArgs of commandsToTest) {
    const out = execFileSync('node', [CLI_PATH, ...cmdArgs], {
      cwd: REPO_ROOT,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe']
    });
    assert.ok(out.length > 20, `Command 'konoha ${cmdArgs.join(' ')}' returned empty output`);
  }
  console.log('✓ CLI help flags passed');

  console.log('\nAll test_commandcode_and_argument_aliases tests passed!');
}

if (require.main === module) {
  run().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { run };
